File: leetcode/cpp/p68-text-justification.cpp
// LeetCode p68. Text Justification.

#include "leetcode.hpp"

#include <algorithm>
#include <cassert>
#include <iostream>
#include <span>
#include <vector>

std::string leftJustify(std::span<std::string const> words, size_t width)
{
    std::string line(width, ' ');
    auto it = line.begin();
    for (const auto& word : words) {
        it = std::copy(word.begin(), word.end(), it);
        ++it;
    }
    return line;
}

std::string leftRightJustify(std::span<std::string const> words, size_t width, size_t lineLen)
{
    if (words.size() < 2) {
        return leftJustify(words, width);
    }
    const size_t totalSpaces = width + words.size() - lineLen;
    const size_t minSpacesBetweenWords = totalSpaces / (words.size() - 1);
    size_t remainderSpacesBetweenWords = totalSpaces % (words.size() - 1);

    std::string line(width, ' ');
    auto it = line.begin();
    for (const auto& word : words) {
        it = std::copy(word.begin(), word.end(), it);
        it += minSpacesBetweenWords;
        if (remainderSpacesBetweenWords > 0) {
            ++it;
            --remainderSpacesBetweenWords;
        }
    }
    return line;
}

class Solution {
public:
    std::vector<std::string> fullJustify(const std::vector<std::string>& words, int maxWidth)
    {
        const size_t width = maxWidth;
        std::vector<std::string> output;
        auto l = words.begin();
        size_t lineLen = 0;
        for (auto r = l; r != words.end(); ++r) {
            const auto& word = *r;
            if (lineLen + word.size() > width) {
                output.push_back(leftRightJustify(std::span(l, r), width, lineLen));
                l = r;
                lineLen = 0;
            }
            lineLen += word.size() + 1;
        }
        if (l != words.end()) {
            output.push_back(leftJustify(std::span(l, words.end()), width));
        }
        return output;
    }
};

void test_leftJustify()
{
    struct TestCase {
        std::vector<std::string> words;
        size_t width;
        std::string exp;
    };

    const TestCase test_cases[] = {
        { { "a" }, 1, "a" },
        { { "a" }, 5, "a    " },
        { { "a", "b" }, 3, "a b" },
        { { "a", "b" }, 4, "a b " },
        { { "a", "b", "c" }, 5, "a b c" },
        { { "a", "b", "c" }, 6, "a b c " },
    };

    for (const auto& tc : test_cases) {
        const auto act = leftJustify(std::span(tc.words), tc.width);
        if (tc.exp != act) {
            std::cerr << "fail. leftJustify(words: "
                      << leetcode::to_string(tc.words)
                      << ", width: " << tc.width
                      << "), exp: '" << tc.exp
                      << "', act: '" << act
                      << "'\n";
        }
    }
}

void test_leftRightJustify()
{
    struct TestCase {
        std::vector<std::string> words;
        size_t width;
        size_t lineLen;
        std::string exp;
    };

    const TestCase test_cases[] = {
        { { "a" }, 1, 2, "a" },
        { { "a" }, 2, 2, "a " },
        { { "a" }, 5, 2, "a    " },
        { { "a", "b" }, 3, 4, "a b" },
        { { "a", "b" }, 4, 4, "a  b" },
        { { "a", "b" }, 5, 4, "a   b" },
        { { "a", "b", "c" }, 5, 6, "a b c" },
        { { "a", "b", "c" }, 6, 6, "a  b c" },
        { { "a", "b", "c" }, 7, 6, "a  b  c" },
        { { "a", "b", "c" }, 8, 6, "a   b  c" },
        { { "This", "is", "an" }, 16, 11, "This    is    an" },
    };

    for (const auto& tc : test_cases) {
        const auto act = leftRightJustify(std::span(tc.words), tc.width, tc.lineLen);
        if (tc.exp != act) {
            std::cerr << "fail. leftJustify(words: "
                      << leetcode::to_string(tc.words)
                      << ", width: " << tc.width
                      << ", lineLen: " << tc.lineLen
                      << "), exp: " << tc.exp
                      << ", act: " << act
                      << "\n";
        }
    }
}

void test_fullJustify()
{
    struct TestCase {
        std::vector<std::string> words;
        int maxWidth;
        std::vector<std::string> exp;
    };

    const TestCase test_cases[] = {
        { { "Imma", "be" }, 7, { "Imma be" } },
        { { "Imma", "be" }, 8, { "Imma be " } },
        { { "Imma", "be", "real" }, 7, { "Imma be", "real   " } },
        { { "Imma", "be", "real" }, 6, { "Imma  ", "be    ", "real  " } },
        { { "Imma", "be", "real" }, 8, { "Imma  be", "real    " } },
        { { "This", "is", "an", "example", "of", "text", "justification." }, 16, { "This    is    an", "example  of text", "justification.  " } },
        { { "What", "must", "be", "acknowledgment", "shall", "be" }, 16, { "What   must   be", "acknowledgment  ", "shall be        " } },
        { { "Science", "is", "what", "we", "understand", "well", "enough", "to", "explain", "to", "a", "computer.", "Art", "is", "everything", "else", "we", "do" }, 20, { "Science  is  what we", "understand      well", "enough to explain to", "a  computer.  Art is", "everything  else  we", "do                  " } },
    };

    Solution solution;
    for (const auto& tc : test_cases) {
        const auto act = solution.fullJustify(tc.words, tc.maxWidth);
        if (tc.exp != act) {
            std::cerr << "fail. leftJustify(words: "
                      << leetcode::to_string(tc.words)
                      << ", width: " << tc.maxWidth
                      << "), exp: " << leetcode::to_string(tc.exp)
                      << ", act: " << leetcode::to_string(act)
                      << "\n";
        }
    }
}

int main()
{
    test_leftJustify();
    test_leftRightJustify();
    test_fullJustify();
    return 0;
}

File: leetcode/rust/src/p68-text-justification.rs
// LeetCode 68. Text Justification.
use std::vec::Vec;

struct Solution {}

fn left_justify(v: &[String], width: usize) -> String {
    return format!("{:<1$}", v.join(" "), width);
}

fn left_right_justify(v: &[String], width: usize, line_len: usize) -> String {
    assert!(v.len() >= 2);

    let total_spaces = (width + v.len()) - line_len;
    let min_spaces_between_words = total_spaces / (v.len() - 1);
    let num_extra_spaces = total_spaces % (v.len() - 1);
    let mut spaces = " ".repeat(min_spaces_between_words + 1);
    let mut line = String::with_capacity(width);
    for (i, s) in v.iter().enumerate() {
        line.push_str(s);
        if i + 1 == v.len() {
            break;
        }
        if i == num_extra_spaces {
            spaces.pop();
        }
        line.push_str(&spaces);
    }
    return line;
}

impl Solution {
    pub fn full_justify(words: Vec<String>, max_width: i32) -> Vec<String> {
        assert!(words.len() >= 1);
        let max_width = max_width as usize;
        let mut l: usize = 0;
        let mut lines: Vec<String> = Vec::new();
        let mut line_len: usize = 0;
        for (r, word) in words.iter().enumerate() {
            let word_len = word.len();
            if line_len + word_len > max_width {
                let is_more_than_one_word = r - l > 1;
                if is_more_than_one_word {
                    lines.push(left_right_justify(&words[l..r], max_width, line_len));
                } else {
                    lines.push(left_justify(&words[l..r], max_width));
                }
                l = r;
                line_len = 0;
            }
            line_len += word_len + 1;
        }
        if l != words.len() {
            lines.push(left_justify(&words[l..], max_width));
        }
        return lines;
    }
}

fn test_left_justify() {
    struct TestCase {
        words: String,
        width: usize,
        exp: String,
    }

    let test_cases = vec![TestCase {
        words: "a".to_string(),
        width: 5,
        exp: "a    ".to_string(),
    }];

    for tc in test_cases.iter() {
        let words: Vec<String> = tc.words.split(" ").map(str::to_string).collect();
        let act = left_justify(&words[..], tc.width);
        if tc.exp != act {
            println!(
                "Fail, words: {:?}, width: {}, exp: {:?}, act: {:?}",
                tc.words, tc.width, tc.exp, act
            );
        }
    }
}

fn test_left_right_justify() {
    struct TestCase {
        words: String,
        width: usize,
        line_len: usize,
        exp: String,
    }

    let test_cases = vec![
        TestCase {
            words: "a b".to_string(),
            width: 3,
            line_len: 4,
            exp: "a b".to_string(),
        },
        TestCase {
            words: "a b".to_string(),
            width: 4,
            line_len: 4,
            exp: "a  b".to_string(),
        },
        TestCase {
            words: "a b c".to_string(),
            width: 5,
            line_len: 6,
            exp: "a b c".to_string(),
        },
        TestCase {
            words: "a b c".to_string(),
            width: 6,
            line_len: 6,
            exp: "a  b c".to_string(),
        },
        TestCase {
            words: "a b c".to_string(),
            width: 7,
            line_len: 6,
            exp: "a  b  c".to_string(),
        },
    ];

    for tc in test_cases.iter() {
        let words: Vec<String> = tc.words.split(" ").map(str::to_string).collect();
        let act = left_right_justify(&words[..], tc.width, tc.line_len);
        if tc.exp != act {
            println!(
                "Fail, words: {:?}, width: {}, line_len: {}, exp: {:?}, act: {:?}",
                tc.words, tc.width, tc.line_len, tc.exp, act
            );
        }
    }
}

fn test_full_justify() {
    struct TestCase {
        words: String,
        max_width: i32,
        exp: Vec<String>,
    }

    let test_cases = vec![
        TestCase {
            words: "Imma be".to_string(),
            max_width: 7,
            exp: vec!["Imma be".to_string()],
        },
        TestCase {
            words: "Imma be".to_string(),
            max_width: 8,
            exp: vec!["Imma be ".to_string()],
        },
        TestCase {
            words: "Imma be real".to_string(),
            max_width: 6,
            exp: vec![
                "Imma  ".to_string(),
                "be    ".to_string(),
                "real  ".to_string(),
            ],
        },
        TestCase {
            words: "Imma be real".to_string(),
            max_width: 7,
            exp: vec!["Imma be".to_string(), "real   ".to_string()],
        },
        TestCase {
            words: "Imma be real".to_string(),
            max_width: 8,
            exp: vec!["Imma  be".to_string(), "real    ".to_string()],
        },
    ];
    for tc in test_cases.iter() {
        let words = tc.words.split(" ").map(str::to_string).collect();
        let act = Solution::full_justify(words, tc.max_width);
        if tc.exp != act {
            println!(
                "Fail. words: {:?}, max_width: {}, exp: {:?}, act: {:?}",
                tc.words, tc.max_width, tc.exp, act
            );
        }
    }
}

fn main() {
    test_left_justify();
    test_left_right_justify();
    test_full_justify();
}
